clippy_lints/src/loops/never_loop.rs

```diff
@@ -263,13 +264,25 @@ fn never_loop_expr<'tcx>(
         | ExprKind::Lit(_)
         | ExprKind::Err(_) => NeverLoopResult::Normal,
     };
-    combine_seq(result, || {
+    let result = combine_seq(result, || {
         if cx.typeck_results().expr_ty(expr).is_never() {
             NeverLoopResult::Diverging
         } else {
             NeverLoopResult::Normal
         }
-    })
+    });
+    if  let NeverLoopResult::Diverging = result &&
+        let Some(macro_call) = root_macro_call_first_node(cx, expr) &&
+        let "todo" | "unimplemented" = cx.tcx.item_name(macro_call.def_id).as_str()
```

Comment:
You can use diagnostic items instead; `todo_macro`/`unimplemented_macro`.

Reply:
I tried that, it didn't work. Could that `sym` be referring to something else?

PS: this is also the method used in the todo restriction lint, so if we can get to the bottom of this we may also consider fixing that code too.

Reply:
Odd, since it's declared here: https://doc.rust-lang.org/nightly/src/core/macros/mod.rs.html#767

Not sure why that wouldn't work then, since `macro_call.def_id` should refer to the `todo!` this is assigned to.

Reply:
```rust
let Some(sym::todo_macro | sym::unimplemented_macro) = cx.tcx.get_diagnostic_name(macro_call.def_id)
```

seems to work fine for me

Reply:
From testing and from reading the documentation, it appears that `sym::todo_macro` represents the string "todo_macro", which... seems useless? And I tried exactly that @Alexendoo , it fails the test on my machine

Reply:
yes, `sym::todo_macro` represents the string "todo_macro", *but* that string represents the diagnostic item `todo!`.

I'll try out the code Alex suggested, I don't see why it wouldn't work.

Reply:
The suggested code works fine for me.

Reply:
Oh, sorry I used `item_name` rather than `get_diagnostic_name`